query/iterator.go

```diff
@@ -404,175 +403,43 @@ func NewCloseInterruptIterator(input Iterator, closing <-chan struct{}) Iterator
 	}
 }
 
-// AuxIterator represents an iterator that can split off separate auxiliary iterators.
-type AuxIterator interface {
-	Iterator
+// IteratorScanner is used to scan the results of an iterator into a map.
+type IteratorScanner interface {
+	// Peek retrieves information about the next point. It returns a timestamp, the name, and the tags.
+	Peek() (int64, string, Tags)
 
-	// Auxilary iterator
-	Iterator(name string, typ influxql.DataType) Iterator
+	// ScanAt will take a time, name, and tags and scan the point that matches those into the map.
+	ScanAt(ts int64, name string, tags Tags, values map[string]interface{})
 
-	// Start starts writing to the created iterators.
-	Start()
+	// Stats returns the IteratorStats from the Iterator.
+	Stats() IteratorStats
+
+	// Err returns an error that was encountered while scanning.
+	Err() error
 
-	// Backgrounds the iterator so that, when start is called, it will
-	// continuously read from the iterator.
-	Background()
+	io.Closer
 }
 
-// NewAuxIterator returns a new instance of AuxIterator.
-func NewAuxIterator(input Iterator, opt IteratorOptions) AuxIterator {
+// SkipDefault is a sentinel value to tell the IteratorScanner to skip setting the
+// default value if none was present. This causes the map to use the previous value
+// if it was previously set.
+var SkipDefault = interface{}(0)
+
+// NewIteratorScanner produces an IteratorScanner for the Iterator.
+func NewIteratorScanner(input Iterator, keys []string, defaultValue interface{}) IteratorScanner {
```

Comment:
Given the reduction in complexity and locking by removing the aux iterator, have you seen any perf improvements as a result of this change?

Reply:
I think we need to test it. I'm not sure how thorough our benchmarks are though.

Reply:
I just pushed a change to reduce memory allocations. There was a regression in memory allocations because the time was being taken care of by eval which resulted in an extra allocation of an interface for every row. Here's the numbers for the one with the memory optimization that avoids the extra allocation:

```
benchmark                            old ns/op      new ns/op     delta
BenchmarkCountIterator_1K-8          42116          43269         +2.74%
BenchmarkCountIterator_100K-8        4191261        4113898       -1.85%
BenchmarkCountIterator_1M-8          41485285       39940193      -3.72%
BenchmarkSampleIterator_1k-8         464            461           -0.65%
BenchmarkSampleIterator_100k-8       410            412           +0.49%
BenchmarkSampleIterator_1M-8         321            306           -4.67%
BenchmarkDistinctIterator_1K-8       74546          72119         -3.26%
BenchmarkDistinctIterator_100K-8     6858535        6741477       -1.71%
BenchmarkDistinctIterator_1M-8       68455510       69030348      +0.84%
BenchmarkModeIterator_1K-8           103948         111675        +7.43%
BenchmarkModeIterator_100K-8         34101326       34786063      +2.01%
BenchmarkModeIterator_1M-8           408322854      232866597     -42.97%
BenchmarkSelect_Raw_1K-8             1724663        197585        -88.54%
BenchmarkSelect_Raw_100K-8           1808846670     192179458     -89.38%
BenchmarkSelect_Dedupe_1K-8          38699956       30059134      -22.33%

benchmark                            old allocs     new allocs     delta
BenchmarkCountIterator_1K-8          11             11             +0.00%
BenchmarkCountIterator_100K-8        11             11             +0.00%
BenchmarkCountIterator_1M-8          11             11             +0.00%
BenchmarkSampleIterator_1k-8         3              3              +0.00%
BenchmarkSampleIterator_100k-8       3              3              +0.00%
BenchmarkSampleIterator_1M-8         3              3              +0.00%
BenchmarkDistinctIterator_1K-8       18             18             +0.00%
BenchmarkDistinctIterator_100K-8     18             18             +0.00%
BenchmarkDistinctIterator_1M-8       18             18             +0.00%
BenchmarkModeIterator_1K-8           22             22             +0.00%
BenchmarkModeIterator_100K-8         42             42             +0.00%
BenchmarkModeIterator_1M-8           53             53             +0.00%
BenchmarkSelect_Raw_1K-8             2068           2091           +1.11%
BenchmarkSelect_Raw_100K-8           2000123        2000096        -0.00%
BenchmarkSelect_Dedupe_1K-8          302008         302032         +0.01%

benchmark                            old bytes     new bytes     delta
BenchmarkCountIterator_1K-8          928           928           +0.00%
BenchmarkCountIterator_100K-8        945           945           +0.00%
BenchmarkCountIterator_1M-8          1101          1032          -6.27%
BenchmarkSampleIterator_1k-8         400           400           +0.00%
BenchmarkSampleIterator_100k-8       400           400           +0.00%
BenchmarkSampleIterator_1M-8         400           400           +0.00%
BenchmarkDistinctIterator_1K-8       7052          7053          +0.01%
BenchmarkDistinctIterator_100K-8     7058          7054          -0.06%
BenchmarkDistinctIterator_1M-8       7032          7032          +0.00%
BenchmarkModeIterator_1K-8           197152        197152        +0.00%
BenchmarkModeIterator_100K-8         44999200      44999200      +0.00%
BenchmarkModeIterator_1M-8           529293856     529293856     +0.00%
BenchmarkSelect_Raw_1K-8             44086         53304         +20.91%
BenchmarkSelect_Raw_100K-8           40010976      48006068      +19.98%
BenchmarkSelect_Dedupe_1K-8          4209411       4210828       +0.03%
```

There's still more memory allocated in general for raw queries for number of bytes used, but it's not as large as the previous change. I'll keep trying to track it down.

CPU usage is generally better across the board with this technique though.